clang/lib/CodeGen/CGDeclCXX.cpp

```diff
@@ -1059,9 +1059,15 @@ CodeGenFunction::GenerateCXXGlobalInitFunc(llvm::Function *Fn,
     if (Guard.isValid()) {
       // If we have a guard variable, check whether we've already performed
       // these initializations. This happens for TLS initialization functions.
-      llvm::Value *GuardVal = Builder.CreateLoad(Guard);
-      llvm::Value *Uninit = Builder.CreateIsNull(GuardVal,
-                                                 "guard.uninitialized");
+      Address GuardAddr = Guard;
+      if (auto *GV = dyn_cast<llvm::GlobalValue>(Guard.getPointer()))
```

Comment:
We need to do this unconditionally, and we should be able to do because `Guard` is a constant address.  But if there's some code pattern where that's not true, we need to fix that code pattern rather than silently just not using this intrinsic.

Reply:
There is a code pattern where this "guarded initialization" is done for non-TLS var ([partitions.cpp test](https://github.com/nextsilicon/next-llvm-project/blob/b36811cc9baf1c72de2fa1c8b5d8fc30bae9a15c/clang/test/CodeGenCXX/partitions.cpp)). That's the reason I've introduced these checks.

Reply:
Then you should pass down whether this is for TLS.

What I'm trying to avoid is some situation where we silently don't emit the intrinsic for a thread local because the guard address we pass in is not directly a `GlobalValue`.

Reply:
I've added changes to pass down `IsTLS` flag from the point where variable is created. That should ensure we always emit TLS intrinsic.